store/store.go

```diff
@@ -2,108 +2,124 @@
+
+	// ErrEndpointsRequired is returned when endpoints are required but none are given
+	ErrEndpointsRequired = errors.New("etcd endpoints for remote etcd cluster required with no-embed")
+	// ErrStoreInitedAlready is returned when the store is already intialized
+	ErrStoreInitedAlready = errors.New("store has been intialized already")
+)
 
 // GDStore is the GlusterD centralized store
 type GDStore struct {
+	conf Config
+
 	*clientv3.Client
 	*concurrency.Session
+
+	ee *elasticetcd.ElasticEtcd
 }
 
-// InitStore creates and initializes the store
-func InitStore() error {
-	address := config.GetString("etcdclientaddress")
-
-	c, e := clientv3.New(clientv3.Config{
-		Endpoints:        []string{address},
-		AutoSyncInterval: 1 * time.Minute,
-		DialTimeout:      10 * time.Second,
-	})
-	if e != nil {
-		log.WithError(e).Error("failed to create etcd client")
-		return e
-	}
-	log.Debug("etcd client connection created")
-
-	// Create a new session (lease kept alive for the lifetime of a client)
-	// This is currently used for:
-	// * distributed locking (Mutex)
-	// * representing liveness of the client
-	s, e := concurrency.NewSession(c, concurrency.WithTTL(sessionTTL))
-	if e != nil {
-		log.WithError(e).Error("failed to create an etcd session")
-		return e
-	}
+// Init initializes the GD2 store
+func Init(conf *Config) error {
+	lock.Lock()
+	defer lock.Unlock()
 
-	// publish liveness of this instance into the store
-	key := livenessKeyPrefix + gdctx.MyUUID.String()
-	if _, e = c.Put(context.TODO(), key, "", clientv3.WithLease(s.Lease())); e != nil {
-		return e
+	if Store != nil {
+		return ErrStoreInitedAlready
 	}
 
-	Store = &GDStore{c, s}
-	return nil
+	var err error
+	Store, err = New(conf)
+	return err
 }
 
-// Close closes the store connections
-func (s *GDStore) Close() {
-	s.Session.Orphan()
-	if e := s.Client.Close(); e != nil {
-		log.WithError(e).Warn("failed to close etcd client connection")
-	}
-	// FIXME: We should close the session first and then the client but it
-	// doesn't work because restart of embedded etcd server when using v3
-	// has issues.
-	if e := s.Session.Close(); e != nil {
-		log.WithError(e).Warn("failed to close etcd session")
-	}
+// Close closes the GD2 store
+func Close() {
+	lock.Lock()
+	defer lock.Unlock()
+
+	Store.Close()
 }
 
-// IsNodeAlive returns true if the node specified is alive as seen by the store
-func (s *GDStore) IsNodeAlive(nodeID interface{}) bool {
+// Destroy closes the GD2 store and deletes the store data dir
+func Destroy() {
+	lock.Lock()
+	defer lock.Unlock()
+
+	Store.Destroy()
+	Store = nil
 
-	var keySuffix string
+	return
+}
 
-	switch nodeID.(type) {
-	case uuid.UUID:
-		keySuffix = nodeID.(uuid.UUID).String()
-	case string:
-		keySuffix = nodeID.(string)
-		if uuid.Parse(keySuffix) == nil {
-			return false
+// New creates a new GDStore from the given Config.
+// If the given Config is nil, the saved store config is used.
+func New(conf *Config) (*GDStore, error) {
```

Comment:
Most of the methods in store can actually be unexported. I'll go through the package again.